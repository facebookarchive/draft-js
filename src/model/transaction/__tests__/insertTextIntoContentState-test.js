/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow strict-local
 * @format
 * @oncall draft_js
 */

'use strict';
import type SelectionState from 'SelectionState';

const CharacterMetadata = require('CharacterMetadata');
const {BOLD} = require('SampleDraftInlineStyle');

const getSampleStateForTesting = require('getSampleStateForTesting');
const insertTextIntoContentState = require('insertTextIntoContentState');

const {contentState, selectionState} = getSampleStateForTesting();

const EMPTY = CharacterMetadata.EMPTY;
const initialBlock = contentState.getBlockMap().first();

const assertInsertTextIntoContentState = (
  text: string | $TEMPORARY$string<'xx'>,
  characterMetadata: $FlowFixMe | CharacterMetadata,
  selection: $FlowFixMe | SelectionState = selectionState,
) => {
  expect(
    insertTextIntoContentState(contentState, selection, text, characterMetadata)
      .getBlockMap()
      .toJS(),
  ).toMatchSnapshot();
};

test('must throw if selection is not collapsed', () => {
  expect(() => {
    insertTextIntoContentState(
      contentState,
      selectionState.set('focusOffset', 2),
      'hey',
      EMPTY,
    );
  }).toThrow();
});

test('must return early if no text is provided', () => {
  assertInsertTextIntoContentState('', EMPTY);
});

test('must insert at the start', () => {
  assertInsertTextIntoContentState(
    'xx',
    CharacterMetadata.create({style: BOLD}),
  );
});

test('must insert within block', () => {
  assertInsertTextIntoContentState(
    'xx',
    CharacterMetadata.create({style: BOLD}),
    selectionState.merge({
      focusOffset: 2,
      anchorOffset: 2,
      isBackward: false,
    }),
  );
});

test('must insert at the end', () => {
  assertInsertTextIntoContentState(
    'xx',
    CharacterMetadata.create({style: BOLD}),
    selectionState.merge({
      focusOffset: initialBlock.getLength(),
      anchorOffset: initialBlock.getLength(),
      isBackward: false,
    }),
  );
});
