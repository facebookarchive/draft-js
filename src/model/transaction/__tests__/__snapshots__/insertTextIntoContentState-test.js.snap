// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`must insert at the end 1`] = `
Object {
  "a": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "a",
    "text": "Alphaxx",
    "type": "unstyled",
  },
  "b": Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  "c": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  "d": Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  "e": Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  "f": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
}
`;

exports[`must insert at the start 1`] = `
Object {
  "a": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "a",
    "text": "xxAlpha",
    "type": "unstyled",
  },
  "b": Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  "c": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  "d": Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  "e": Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  "f": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
}
`;

exports[`must insert within block 1`] = `
Object {
  "a": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "a",
    "text": "Alxxpha",
    "type": "unstyled",
  },
  "b": Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  "c": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  "d": Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  "e": Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  "f": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
}
`;

exports[`must return early if no text is provided 1`] = `
Object {
  "a": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "a",
    "text": "Alpha",
    "type": "unstyled",
  },
  "b": Object {
    "characterList": Array [
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
      Object {
        "entity": "2",
        "style": Array [
          "BOLD",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "b",
    "text": "Bravo",
    "type": "unordered-list-item",
  },
  "c": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
      Object {
        "entity": null,
        "style": Array [],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "c",
    "text": "Test",
    "type": "code-block",
  },
  "d": Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "d",
    "text": "",
    "type": "code-block",
  },
  "e": Object {
    "characterList": Array [],
    "data": Object {},
    "depth": 0,
    "key": "e",
    "text": "",
    "type": "code-block",
  },
  "f": Object {
    "characterList": Array [
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
      Object {
        "entity": null,
        "style": Array [
          "ITALIC",
        ],
      },
    ],
    "data": Object {},
    "depth": 0,
    "key": "f",
    "text": "Charlie",
    "type": "blockquote",
  },
}
`;
